<a href="https://colab.research.google.com/github/max-moore/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS18_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [ ]:
# YOUR WORK HERE

import numpy as np
import pandas as pd
from scipy import stats

In [ ]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-16 16:05:23--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.03s   

2020-07-16 16:05:24 (712 KB/s) - ‘student.zip’ saved [20478/20478]



In [ ]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [ ]:
df = pd.read_csv('student-mat.csv', sep=';')

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [ ]:
# YOUR WORK HERE

# Creating contingency tables

observed = pd.crosstab(df['Dalc'],df['failures'])
observed2 = pd.crosstab(df['famrel'], df['absences'])
observed3 = pd.crosstab(df['famsize'], df['address'])
observed4 = pd.crosstab(df['goout'], df['failures'])

In [ ]:
# Assigning and printing chi2 value, p value, depth of freedom, 
# and expected frequencies

chi2, p_value, dof, expected = stats.chi2_contingency(observed)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

Testing Daytime Alcohol Consumption vs Class Failures

1. Null Hypothesis:
Workday alcohol consumption and class failures are INDEPENDENT of each other, they are NOT statistically correlated.
2. Alternative Hypothesis:
Workday alcohol consumption and class failures are DEPENDENT of each other, they ARE statistically correlated.
3. Confidence Level:
95%

Conclusion:
Based on a chi^2 statistic of 22.6 and a p-value of 0.031, **we reject the null hypothesis** that workday alcohol consumption and class failures are independent and suggest the alternative that there is a strong relationship between these two categorical variables.

This follows intuition - most people would expect frequent workday alcohol consumption to result in worse class performance

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed2)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

Testing Quality of Family Relationships vs Class Absences
1. Null Hypothesis:
Quality of family relationships and class absences are INDEPENDENT of each other, they are NOT statistically correlated.
2. Alternative Hypothesis:
Quality of family relationships and class absences are DEPENDENT of each other, they ARE statistically correlated.
3. Confidence Level:
95%

Conclusion:
Based on a chi^2 statistic of 102.1 and a p-value of 0.974, **we fail to reject the null hypothesis** that quality of family relationships and class absences are independent.

To say this follows intuition or not is difficult.  Students suffering at home from poor family relationships might stay at home because school is a stress compounder - or they might attend school frequently to get away from home.

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed3)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 1.7080569422911414
p value 0.19123766649306323
degrees of freedom 1
expected frequencies table 
 [[ 62.60253165 218.39746835]
 [ 25.39746835  88.60253165]]


Testing Family Size vs Address (Urban or Rural)
1. Null Hypothesis:
Family size and address are INDEPENDENT of each other, they are NOT statistically correlated.
2. Alternative Hypothesis:
Family size and address are DEPENDENT of each other, they ARE statistically correlated.
3. Confidence Level:
95%

Conclusion:
Based on a chi^2 statistic of 1.7 and a p-value of 0.191, **we fail to reject the null hypothesis** that family size and address are independent.

I could imagine these variables having a dependent relationship but am not entirely surprised that the results are not statistically significant - where you live shouldn't have the biggest impact on how many kids you want to have.

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed4)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 28.962714739524678
p value 0.003990520196627485
degrees of freedom 12
expected frequencies table 
 [[ 18.16708861   2.91139241   0.98987342   0.93164557]
 [ 81.35696203  13.03797468   4.43291139   4.1721519 ]
 [102.6835443   16.4556962    5.59493671   5.26582278]
 [ 67.92911392  10.88607595   3.70126582   3.4835443 ]
 [ 41.86329114   6.70886076   2.28101266   2.14683544]]


Testing Time Spent out with Friends vs Class Failures
1. Null Hypothesis:
Time spent out and class failures are INDEPENDENT of each other, they are NOT statistically correlated.
2. Alternative Hypothesis:
Time spent out and class failures are DEPENDENT of each other, they ARE statistically correlated.
3. Confidence Level:
95%

Conclusion:
Based on a chi^2 statistic of 28.9 and a p-value of 0.003, **we reject the null hypothesis** that time spent out and class failures are independent.

This follows intuition.  Time spent out with friends is time that isn't spent improving school performance.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
# YOUR WORK HERE

# Contingency table

contingency = pd.crosstab(df['Dalc'],df['failures'], margins=True)

contingency

failures    0   1   2   3  All
Dalc                          
1         232  29   7   8  276
2          50  12   8   5   75
3          17   6   1   2   26
4           6   2   1   0    9
5           7   1   0   1    9
All       312  50  17  16  395

In [ ]:
# Sorting row totals

row_totals = df['failures'].value_counts().sort_index().values

row_totals

array([312,  50,  17,  16])

In [ ]:
# Sorting column totals

col_totals = df['Dalc'].value_counts().sort_index().values

col_totals

array([276,  75,  26,   9,   9])

In [ ]:
# Amount of observations (?)

sample_size = df.shape[0]

sample_size

395

In [ ]:
# Template array for expected values

expected = np.array([[0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],
                     [0.0, 0.0, 0.0, 0.0],])

# Assigning values to expected values array

for row_position, col in enumerate(col_totals):
  for column_position, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[row_position][column_position] = expected_value

expected

array([[218.00506329,  34.93670886,  11.87848101,  11.17974684],
       [ 59.24050633,   9.49367089,   3.2278481 ,   3.03797468],
       [ 20.53670886,   3.29113924,   1.11898734,   1.05316456],
       [  7.10886076,   1.13924051,   0.38734177,   0.36455696],
       [  7.10886076,   1.13924051,   0.38734177,   0.36455696]])

In [ ]:
# Obtaining Chi2 statistic

chi2 = ((observed - expected)**2 / expected).sum().sum()

chi2

22.613784416016074

In [ ]:
# Obtaining Degrees of Freedom

dof = (row_totals.size - 1) * (col_totals.size - 1)

dof

12

In [ ]:
# Used the University of Iowa Applet to get p value from chi2

p_value = 0.03119

In [ ]:
# Print everything we need to see

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 22.613784416016074
p value 0.03119
degrees of freedom 12
expected frequencies table 
 [[218.00506329  34.93670886  11.87848101  11.17974684]
 [ 59.24050633   9.49367089   3.2278481    3.03797468]
 [ 20.53670886   3.29113924   1.11898734   1.05316456]
 [  7.10886076   1.13924051   0.38734177   0.36455696]
 [  7.10886076   1.13924051   0.38734177   0.36455696]]


## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)